core/chains/starknet/chain.go

```diff
@@ -2,57 +2,121 @@ package starknet
 	"github.com/smartcontractkit/chainlink-starknet/relayer/pkg/starknet"
-	"github.com/smartcontractkit/chainlink-starknet/relayer/pkg/starknet/db"
 
 	"github.com/smartcontractkit/sqlx"
 
 	"github.com/smartcontractkit/chainlink/core/chains/starknet/types"
 	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/keystore"
 	"github.com/smartcontractkit/chainlink/core/utils"
 )
 
-var _ starknet.Chain = (*chain)(nil)
+var _ starkChain.Chain = (*chain)(nil)
 
 type chain struct {
 	utils.StartStopOnce
 	id   string
-	cfg  starknet.Config
+	cfg  config.Config
 	orm  types.ORM
 	lggr logger.Logger
+	txm  txm.StarkTXM
 }
 
-func NewChain(db *sqlx.DB, dbchain types.DBChain, orm types.ORM, lggr logger.Logger) (*chain, error) {
-	cfg := starknet.NewConfig(*dbchain.Cfg, lggr)
+func NewChain(db *sqlx.DB, ks keystore.StarkNet, dbchain types.DBChain, orm types.ORM, lggr logger.Logger) (ch *chain, err error) {
+	cfg := config.NewConfig(*dbchain.Cfg, lggr)
 	lggr = lggr.With("starknetChainID", dbchain.ID)
-	var ch = chain{
+
+	ch = &chain{
 		id:   dbchain.ID,
 		cfg:  cfg,
 		orm:  orm,
 		lggr: lggr.Named("Chain"),
 	}
 
-	return &ch, nil
+	getClient := func() (caigotypes.Provider, error) {
+		return ch.getClient()
+	}
+
+	ch.txm, err = txm.New(lggr, ks, cfg, getClient)
+	if err != nil {
+		return nil, err
+	}
+
+	return ch, nil
 }
 
-func (c *chain) Config() starknet.Config {
+func (c *chain) Config() config.Config {
 	return c.cfg
 }
 
 func (c *chain) UpdateConfig(cfg *db.ChainCfg) {
 	c.cfg.Update(*cfg)
 }
 
+func (c *chain) TxManager() txm.TxManager {
+	return nil
+}
+
+func (c *chain) Reader() (starknet.Reader, error) {
+	return nil, nil
+}
+
+// getClient returns a client, randomly selecting one from available and valid nodes
+func (c *chain) getClient() (*starknet.Client, error) {
+	var node db.Node
+	var client *starknet.Client
+	nodes, cnt, err := c.orm.NodesForChain(c.id, 0, math.MaxInt)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get nodes")
+	}
+	if cnt == 0 {
+		return nil, errors.New("no nodes available")
+	}
+	rand.Seed(time.Now().Unix()) // seed randomness otherwise it will return the same each time
+	// #nosec
+	index := rand.Perm(len(nodes)) // list of node indexes to try
+	timeout := c.cfg.RequestTimeout()
+	for _, i := range index {
+		node = nodes[i]
+		// create client and check
+		client, err = starknet.NewClient(node.ChainID, node.URL, c.lggr, &timeout)
+		// if error, try another node
+		if err != nil {
+			c.lggr.Warnw("failed to create node", "name", node.Name, "solana-url", node.URL, "error", err.Error())
+			continue
+		}
+		// if all checks passed, mark found and break loop
+		break
+	}
+	// if no valid node found, exit with error
+	if client == nil {
+		return nil, errors.New("no node valid nodes available")
+	}
+	c.lggr.Debugw("Created client", "name", node.Name, "solana-url", node.URL)
+	return client, nil
+}
```

Comment:
Didn't we decide to move client construction into the relay?

Reply:
we could do that and just pass a set of `nodes` or a random `node` back 🤔 - this just follows the same model as solana/terra where the client is constructed alongside the `Chain` since it requires the ORM for retrieving configs

Reply:
Yeah, this is "generic" code we repeat for every integration.

Ideally, CL core should expose the ability to get node (random, or by name/id) and the relayer module should expose the ability to construct a reader/writer/client for a specific node.

Reply:
There is some "solana" stuff left in the debug messages heh
@aalu1418 could you please elaborate more on why you construct the client in the core instead of relayer?
We pass the node config to the relayer..

Reply:
it was mainly to follow the previous code from solana and terra to get it working. I'm open to changing this, but it might be better suited in a separate PR :)